MNT Add specific error message to `get_tags()` when users pass estimator class instead of instance

AnneBeyer opened the pull request:
#### Reference Issues/PRs
Related to [this comment](https://github.com/scikit-learn/scikit-learn/issues/32394#issuecomment-3375647854) in #32394 

#### What does this implement/fix? Explain your changes.
Adds a more specific error message (and respective tests) for cases when an estimator class is passed instead of an instance. This is raised in `get_tags()` and will be thrown for example in `is_regressor()` and all related checks based on `get_tags()`  (taking up the attempt from #32428, as suggested by @StefanieSenger)

#### Any other comments?
I got a `line too long` error when checking the error message, so I moved it to a variable, but I'm not sure if this is the most concise solution here. Feedback welcome!

AnneBeyer committed:
add specific error message when users pass estimator class instead of instance to `is_regressor()` and co.

AnneBeyer committed:
add test for new error message

StefanieSenger commented:
Hi, thanks for your PR, @AnneBeyer 

I didn't review yet, but you need to add a changelog entry (mentioning both the author of #32428 and yourself).

On the variable name for a string: yes  :+1: that's definitely better than `if "long string" in "another long string"`
Apart from that it helps to have a ruler in your VSCode that indicates 88 characters and use the "rewrap" extention to format locally. You can also sometimes  use `\`.

AnneBeyer committed:
add changelog entry

StefanieSenger commented on doc/whats_new/upcoming_changes/sklearn.base/32565.enhancement.rst:
```diff
@@ -0,0 +1,4 @@
+-:func:`base.is_regressor`, :func:`base.is_classifier`, and :func:`base.is_clusterer`
+  now provide helpful error messages when users accidentally pass a class instead
+  of an instance.
+  By :user:`Achyuthan S <Achyuthan-S>` and :user:`Anne Beyer <AnneBeyer>`
```
```suggestion
-:func:`base.is_regressor`, :func:`base.is_classifier`, :func:`base.is_clusterer`
  and :func:`is_outlier_detector` now provide helpful error messages when users 
  accidentally pass a class instead of an instance.
  By :user:`Achyuthan S <Achyuthan-S>` and :user:`Anne Beyer <AnneBeyer>`.
```
`is_outlier_detector` as well, no?

StefanieSenger commented:
You should also adjust your PR title to indicate what type of change you want to merge. In this case I would say it is either FIX or MNT, not so sure myself (and it doesn't matter so much).

> I added the changelog entry under sklearn.base, which is where the functions potentially throwing the error live, but the error message is actually defined in sklearn.utils. Let me know if this is correct.

Yes, that's correct. The changlog is sorted by API, so where the changed function or class officially lives is the right place.

StefanieSenger reviewed (commented):
Thanks for taking over the PR, @AnneBeyer!

Adding the error handling in `get_tags()` increases the scope beyond the `is_classifier`, etc... functions, because [meta-estimators](https://scikit-learn.org/stable/glossary.html#term-meta-estimator) also take `estimator` as an argument. In this case, we need to adjust the test a bit and make sure we don't have the same logic (check for class instead of object) in other places of the code. I think it would be neat! :)

I have also described alternatives for adding the error via param validation in the [comment](https://github.com/scikit-learn/scikit-learn/pull/32565/files#r2461240910).

Also the requested docstring changes can happen in the same PR.

AnneBeyer committed:
add `is_outlier_detector` to changelog

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

AnneBeyer commented on sklearn/tests/test_base.py:
```diff
@@ -334,6 +334,15 @@ def test_is_clusterer(estimator, expected_result):
 
+@pytest.mark.parametrize(
+    "is_func, estimator",
+    [(is_regressor, SVR), (is_classifier, SVC), (is_clusterer, KMeans)],
```
I missed it because there are no tests for `is_outlier_detector` (at least in this file), as opposed to the other three. Should I add those as well (in a separate PR)?

AnneBeyer commented:
Side note: The docstrings were already adapted in #32415.

AnneBeyer committed:
change `is_func` to `is_type` and  add `is_outlier_detector` and meta-estimators to tests

AnneBeyer committed:
Merge branch 'error_message-is_' of https://github.com/AnneBeyer/scikit-learn into error_message-is_

AnneBeyer committed:
update test and changelog to better reflect scope

AnneBeyer commented:
Ok, I think I understand the scope better now:

The error is raised whenever `get_tags()` is called with a class instead of an object, so it is not directly related to the `is_type` functions, but rather checks this **whenever** the tags of an estimator are accessed in any way. 
(See the note above on the `cross_validate` example, where the message is much clearer than
```
TypeError: ElasticNet.__sklearn_tags__() missing 1 required positional argument: 'self'
``` 
was before).

I updated the changelog entry and added a test case in `sklearn.utils.test.test_tags.py` to reflect this better, and I think this makes the tests in `sklearn/tests/test_base.py` redundant.

StefanieSenger reviewed (commented):
Thanks @AnneBeyer! That now looks pretty nice. I agree that the test in sklearn/tests/test_base.py is now redundant and we should remove it.

Apart from that this PR looks good to me.

AnneBeyer committed:
remove redundant tests

AnneBeyer committed:
Update sklearn/utils/_tags.py

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

AnneBeyer committed:
follow multi-line docstring convention

AnneBeyer committed:
Merge branch 'error_message-is_' of https://github.com/AnneBeyer/scikit-learn into error_message-is_

AnneBeyer committed:
remove left-over closing bracket

AnneBeyer committed:
Merge branch 'main' into error_message-is_

StefanieSenger reviewed (approved):
That looks good to me, approving. :)

AnneBeyer committed:
Apply suggestion

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

ogrisel reviewed (commented):
Thanks for the PR, here is some more feedback.

AnneBeyer committed:
change wording of error message

AnneBeyer committed:
Merge branch 'error_message-is_' of https://github.com/AnneBeyer/scikit-learn into error_message-is_

AnneBeyer committed:
check for type instead of error message

AnneBeyer committed:
align error message with #32888

GaelVaroquaux reviewed (approved):
LGTM

I made a cosmetic comment, but it should not prevent merge

StefanieSenger reviewed (commented):
> Isn't it the other way around?

Haha, true :sweat_smile:: `class` is a reserved keyword and `cls` is not. But it is used by convention for something very specific, which is not variable naming. And that can cause wrong expectations / confusion for people reading this code later. (You never know which complexity may be added later or where this test is copy pasted to.)

I have seen we have been using `klass` in other parts of the code. I think it is good to change `est` into something else, please pick an option and finish this as you like, @AnneBeyer. After this we can merge.

The changed approach still looks fine.

AnneBeyer committed:
rename test variable
